hazelcast-jet-s3/src/main/java/com/hazelcast/jet/s3/S3Sources.java

```diff
@@ -0,0 +1,222 @@
+                List<String> bucketNames,
+                String prefix,
+                String charsetName,
+                Context context,
+                SupplierEx<? extends AmazonS3> clientSupplier,
+                BiFunctionEx<String, String, ? extends T> mapFn
+        ) {
+            this.amazonS3 = clientSupplier.get();
+            this.mapFn = mapFn;
+            this.charset = Charset.forName(charsetName);
+            this.processorIndex = context.globalProcessorIndex();
+            this.totalParallelism = context.totalParallelism();
+            this.listingMap = bucketNames
+                    .stream()
+                    .collect(toMap(key -> key, key -> amazonS3.listObjects(key, prefix)));
+        }
+
+        private void fillBuffer(SourceBuffer<? super T> buffer) throws IOException {
+            if (reader != null) {
+                addBatchToBuffer(buffer);
+                return;
+            }
+
+            if (iterator == null) {
+                // create an iterator for the object summaries which belongs to
+                // this processor.
+                iterator = createIterator();
+                // iterator is empty, we've exhausted all the objects
+                if (!iterator.hasNext()) {
+                    buffer.close();
+                    return;
+                }
+            }
+
+            if (iterator.hasNext()) {
+                S3ObjectSummary summary = iterator.next();
+                S3Object s3Object = amazonS3.getObject(summary.getBucketName(), summary.getKey());
+                objectName = s3Object.getKey();
+                reader = new BufferedReader(new InputStreamReader(s3Object.getObjectContent(), charset));
+                addBatchToBuffer(buffer);
+            } else {
+                iterator = null;
+                listingMap.replaceAll((k, v) -> v.isTruncated() ? amazonS3.listNextBatchOfObjects(v) : EMPTY_LISTING);
+            }
+        }
+
+        private void addBatchToBuffer(SourceBuffer<? super T> buffer) throws IOException {
+            for (int i = 0; i < BATCH_COUNT; i++) {
+                String line = reader.readLine();
+                if (line == null) {
+                    reader.close();
+                    reader = null;
+                    return;
+                }
+                buffer.add(mapFn.apply(objectName, line));
+            }
+        }
+
+        private Iterator<S3ObjectSummary> createIterator() {
+            return listingMap
+                    .values()
+                    .stream()
+                    .flatMap(listing -> listing.getObjectSummaries().stream())
+                    .filter(summary -> belongsToThisProcessor(summary.getKey()))
+                    .iterator();
+        }
+
+
+        private boolean belongsToThisProcessor(String key) {
+            return Math.floorMod(key.hashCode(), totalParallelism) == processorIndex;
+        }
+
+        private void close() {
+            amazonS3.shutdown();
```

Comment:
any open readers (if there was some exception, or job is cancelled) should also be closed here